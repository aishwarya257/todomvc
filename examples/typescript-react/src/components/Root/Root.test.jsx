import {screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Root from './Root';
import {renderWithRouter} from '../../setUpTests';

const setup = (path = '/') => {
    const testData = [
        {
            id: 1,
            completed: false,
            title: 'Title 1',
            badges: ['Hey']
        },
        {
            id: 2,
            completed: true,
            title: 'Title 2',
            badges: ['Hi']
        }
    ];
    window.localStorage.setItem('react-todos', JSON.stringify(testData));
    const data = renderWithRouter(<Root />, {route: path});
    return {
        testData,
        ...data
    };
};

const runActiveTest = (testData) => {
    expect(screen.queryByText(testData[1].title)).not.toBeInTheDocument();
    expect(screen.queryByText(testData[0].title)).toBeInTheDocument();
};

const runCompletedTest = (testData) => {
    expect(screen.queryByText(testData[1].title)).toBeInTheDocument();
    expect(screen.queryByText(testData[0].title)).not.toBeInTheDocument();
};

const runAllTest = (testData) => {
    expect(screen.queryByText(testData[1].title)).toBeInTheDocument();
    expect(screen.queryByText(testData[0].title)).toBeInTheDocument();
};

const buttonClickTest = ({name, fn}) => {
    const {testData} = setup('/');
    runAllTest(testData);
    const activeButton = screen.getByText(name);
    userEvent.click(activeButton);
    fn(testData);
};

describe('test Root component', () => {
    it('Snapshot test', () => {
        const {container} = setup('/');
        expect(container.firstChild).toMatchSnapshot();
    });
    it('Change from show all to active', () => {
        const {testData} = setup('/active');
        runActiveTest(testData);
    });
    it('change from active to completed', () => {
        const {testData} = setup('/completed');
        runCompletedTest(testData);
    });

    it('Testing Active button', () => buttonClickTest({name: 'Active', fn: runActiveTest}));

    it('Testing Completed Button', () =>
        buttonClickTest({name: 'Completed', fn: runCompletedTest}));

    it('Testing All Button', () => buttonClickTest({name: 'All', fn: runAllTest}));

    it('Mark all complete', () => {
        setup();
        const checkbox = screen.getByLabelText('Mark all as complete');
        expect(
            screen.getByText((content, node) => node.textContent === '1 item left')
        ).toBeInTheDocument();
        userEvent.click(checkbox);
        expect(
            screen.getByText((content, node) => node.textContent === '0 left')
        ).toBeInTheDocument();
    });

    it('Clear completed', () => {
        const {testData} = setup();
        const clearCompletedButton = screen.getByText(/Clear Completed/i);
        userEvent.click(clearCompletedButton);
        runActiveTest(testData);
    });
});
