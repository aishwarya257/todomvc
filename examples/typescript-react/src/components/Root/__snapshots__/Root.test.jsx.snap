// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Root component Snapshot test 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      todos
    </h1>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
      value=""
    />
    <section
      class="main"
    >
      <input
        checked=""
        class="toggle-all"
        id="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
    </section>
  </header>
  <ul
    class="todo-list"
  >
    <li
      class=""
    >
      <div
        class="view"
      >
        <input
          class="toggle"
          type="checkbox"
        />
        <label>
          <span>
             
            Title 1
             
          </span>
          <span
            class="badge"
            title="Hey"
          >
            Hey
          </span>
        </label>
        <button
          aria-label="remove"
          class="destroy"
        >
           
           
        </button>
      </div>
    </li>
    <li
      class="completed"
    >
      <div
        class="view"
      >
        <input
          checked=""
          class="toggle"
          type="checkbox"
        />
        <label>
          <span>
             
            Title 2
             
          </span>
          <span
            class="badge"
            title="Hi"
          >
            Hi
          </span>
        </label>
        <button
          aria-label="remove"
          class="destroy"
        >
           
           
        </button>
      </div>
    </li>
  </ul>
  <footer
    class="footer"
  >
    <span
      class="todo-count"
    >
      <strong>
        1
      </strong>
       item left
    </span>
    <ul
      class="filters"
    >
      <li>
        <a
          class="selected"
          href="/"
        >
          All
        </a>
      </li>
      <li>
        <a
          class=""
          href="/active"
        >
          Active
        </a>
      </li>
      <li>
        <a
          class=""
          href="/completed"
        >
          Completed
        </a>
      </li>
    </ul>
    <button
      class="clear-completed"
    >
       
      Clear completed
       
    </button>
  </footer>
</div>
`;
